tower-balance/src/lib.rs

```diff
@@ -283,3 +299,137 @@ impl<F: Future, E> Future for ResponseFuture<F, E> {
+
+    impl Discover for ReluctantDisco {
+        type Key = usize;
+        type Request = ();
+        type Response = ();
+        type Error = ();
+        type Service = ReluctantService;
+        type DiscoverError = ();
+
+        fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::DiscoverError> {
+            let r = self.0.pop_front().map(Async::Ready).unwrap_or(Async::NotReady);
+            debug!("polling disco: {:?}", r.is_ready());
+            Ok(r)
+        }
+    }
+
+    impl Service for ReluctantService {
+        type Request = ();
+        type Response = ();
+        type Error = ();
+        type Future = future::FutureResult<(), ()>;
+
+        fn poll_ready(&mut self) -> Poll<(), ()> {
+            if self.polls_until_ready == 0 {
+                return Ok(Async::Ready(()));
+            }
+
+            self.polls_until_ready -= 1;
+            return Ok(Async::NotReady);
+        }
+
+        fn call(&mut self, _: ()) -> Self::Future {
+            future::ok(())
+        }
+    }
+
+    quickcheck! {
+        /// Creates a random number of services, each of which must be polled a random
+        /// number of times before becoming ready. As the balancer is polled, ensure that
+        /// it does not become ready prematurely and that services are promoted from
+        /// not_ready to ready.
+        fn poll_ready(service_tries: Vec<usize>) -> TestResult {
+            // Stores the number of pending services after each poll_ready call.
+            let mut pending_at = Vec::new();
+
+            let disco = {
+                let mut changes = VecDeque::new();
+                for (i, n) in service_tries.iter().map(|n| *n).enumerate() {
+                    for j in 0..n {
+                        if j == pending_at.len() {
+                            pending_at.push(1);
+                        } else {
+                            pending_at[j] += 1;
+                        }
+                    }
+
+                    let s = ReluctantService { polls_until_ready: n };
+                    changes.push_back(Change::Insert(i, s));
+                }
+                ReluctantDisco(changes)
+            };
+            pending_at.push(0);
+
+            let mut balancer = Balance::new(disco, choose::RoundRobin::default());
+
+            let services = service_tries.len();
+            let mut next_pos = 0;
+            for pending in pending_at.iter().map(|p| *p) {
+                assert!(pending <= services);
+                let ready = services - pending;
+
+                match balancer.poll_ready() {
+                    Err(_) => return TestResult::error("poll_ready failed"),
+                    Ok(p) => {
+                        if p.is_ready() != (ready > 0) {
+                            return TestResult::failed();
+                        }
+                    }
+                }
+
+                if balancer.len() != ready {
+                    return TestResult::failed();
```

Comment:
Here you make a bunch of boolean assertions and fail the test if any of them is true. If this test fails, I think it would be nice to know *which* assertion was the one that failed, to aid in debugging any regressions. 

I'd consider using `TestResult::error`, which AFAIK is the only way to add a message to a QuickCheck test failure. I know it's technically supposed to represent a test failure "due to a runtime error", but my understanding is that this doesn't actually change anything except for the message that QuickCheck will print on failure (let me know if I'm incorrect!).

Reply:
Alternatively, I suppose you could replace these with `assert!()`, though I'm not sure how that interacts with QuickCheck off the top of my head --- will a panic stop quickcheck from shrinking etc, or is it just treated as a failure?

Reply:
Another option is to consider breaking this test into multiple tests that each only check for a _specific_ property --- my understanding is that this is considered a more idiomatic pattern in property-based testing. Of course if the test already takes a long time to run, this may not be ideal either.

Reply:
I agree that narrower tests are desirable. This is better than what we had.  We should add more tests.

Reply:
fwiw, `error` is meant to indicate runtime errors. `failed` is the "right" thing to use here.